tfjs-backend-webgpu/src/binary_op_webgpu.ts

```diff
@@ -50,46 +49,54 @@ export class BinaryOpProgram implements WebGPUProgram {
         bShape.length <= 1 && aShape.length > 1 && bShape[0] < 128;
 
     if (this.useSharedMemoryWithA || this.useSharedMemoryWithB) {
-      this.isVec4 = false;
+      this.outputComponent = 1;
       // lastDimensionSize is used as sharedBuf array size, so can not be
       // used as uniform.
       this.lastDimensionSize =
           this.useSharedMemoryWithB ? bShape[0] : aShape[0];
-      this.shaderKey = `binary_${this.type}_${op}_${this.lastDimensionSize}_${
-          this.useSharedMemoryWithB}`;
+      this.shaderKey = `binary_${this.type}_${op}_${this.lastDimensionSize}`;
       this.type = 'shared';
       // This is an experimental value when using shared memory.
       // Note that the maximum of workgroup X dimension is 256.
       this.workgroupSize = [256, 1, 1];
-      this.workPerThread = 1;
     } else {
-      if (util.arraysEqual(aShape, bShape) &&
-          util.sizeFromShape(aShape) % 4 === 0) {
-        this.isVec4 = true;
+      const aDivisibleBy4 =
+          aShape.length > 0 && aShape[aShape.length - 1] % 4 === 0;
+      const bDivisibleBy4 =
+          bShape.length > 0 && bShape[bShape.length - 1] % 4 === 0;
+      if (aDivisibleBy4 && bDivisibleBy4) {
         this.outputComponent = 4;
         this.type = 'vec4';
-        this.workPerThread = 4;
+      } else if (
+          (op === BinaryOpType.SUB || op === BinaryOpType.ADD ||
+           op === BinaryOpType.MUL || op === BinaryOpType.DIV) &&
+          ((aDivisibleBy4 &&
+            (util.isScalarShape(bShape) || bShape[bShape.length - 1] === 1)) ||
+           (bDivisibleBy4 &&
+            (util.isScalarShape(aShape) || aShape[aShape.length - 1] === 1)))) {
+        this.type = 'custom';
+        this.outputComponent = 4;
+        this.variableComponents = aDivisibleBy4 ? [4, 1] : [1, 4];
       } else {
-        this.isVec4 = false;
+        this.outputComponent = 1;
         this.type = 'plain';
-        this.workPerThread = 1;
       }
-      this.shaderKey = `binary_${this.type}_${op}`;
+      this.shaderKey = `binary_${this.type}_${op}_${this.variableComponents}`;
```

Comment:
I still hope variableComponents has a meaningful value for all the paths instead of undefined.